### ISSUE TITLE
Feature idea: typespec diffs / dialyzer diffs

### ISSUE THREAD

**Issue by Nezteb** [zestcreative/utility#25]
This is probably out of scope but I've been wrangling some `dialyzer` errors lately, and it's a huge pain to have a type like this:

```elixir
@type t :: %IdpData{
        id: binary(),
        sp_id: binary(),
        base_url: nil | binary(),
        metadata_file: nil | binary(),
        use_redirect_for_requests: boolean(),
        sp_sign_requests: boolean(),
        sp_sign_metadata: boolean(),
        idp_signs_assertions: boolean(),
        idp_signs_envelopes: boolean(),
        allow_idp_initiated_flow: boolean(),
        allowed_target_urls: nil | [binary()],
        entity_id: binary(),
        signed_requests: binary(),
        certificates: certificates(),
        sso_redirect_url: url(),
        sso_post_url: url(),
        slo_redirect_url: url(),
        slo_post_url: url(),
        name_id_format: name_id_format(),
        fingerprints: [binary()],
        esaml_idp_record: map(),
        esaml_sp_record: map()
      }
```

And get a vague dialyzer error like this:
```
Contract cannot be correct because return type on line number 154 is mismatched.

Success typing (line 154):
%IdpData{
  :allow_idp_initiated_flow => false | nil | true,
  :allowed_target_urls => _,
  :base_url => _,
  :certificates => nil | [any()],
  :entity_id => nil | binary(),
  :esaml_idp_record =>
    nil
    | {:esaml_idp_metadata, {_, _, _, _}, {_, _, _}, true, :undefined, [any()], [any()],
       :undefined, :unknown | [any()]},
  :esaml_sp_record =>
    nil
    | {:esaml_sp, {_, _, _, _}, {_, _, _}, _, _, [], false | nil | true, false | nil | true,
       false | nil | true, true, false | nil | true, nil | [any()], nil | [any()],
       nil | [any()], :undefined, false, :undefined | [any()]},
  :fingerprints => nil | [binary()],
  :id => _,
  :idp_signs_assertions => false | nil | true,
  :idp_signs_envelopes => false | nil | true,
  :metadata_file => _,
  :name_id_format => :unknown | string(),
  :signed_requests => nil | binary(),
  :slo_post_url => nil | :unknown | binary(),
  :slo_redirect_url => nil | :unknown | binary(),
  :sp_id => _,
  :sp_sign_metadata => false | nil | true,
  :sp_sign_requests => false | nil | true,
  :sso_post_url => nil | :unknown | binary(),
  :sso_redirect_url => nil | :unknown | binary(),
  :use_redirect_for_requests => false | nil | true,
  :valid? => boolean()
}
```

Again, this is hyper-specific to `dialyzer` and typespecs, so it might not be relevant to `elixirstream`.

**Comment by dbernheisel** [comment]
not sure I'm following completely. Can you help me understand with a fuller example?

Thinking about some steps to find what (i think) you're asking about, using Phoenix as an example:

1. Generate a project using older Phoenix.
2. Upgrade the project's Phoenix version
3. Run dialyzer and save the warnings

Is that what you're looking for?

**Comment by Nezteb** [comment]
Yeah, I did a terrible job explaining. It'd actually be completely unrelated to the purpose of your original diff generator. Literally the only similarity is that it'd be a diff tool for Elixir code. 😂

I guess a BDD description of the tool I'm imagining is: Given two typespecs or two structs, when each is potentially complex and deeply nested, when the two are not equivalent, then produce a highlighted diff.

I guess I could just encode each as JSON and use any old JSON diffing tool. 🤔

I'll just close this for now; I just spent a few hours struggling with obscure Dialyzer errors on a new project and was trying to come up with an idea for a tool that could alleviate the pain. My first thought was an in-browser tool that could intelligently diff Elixir code, which is why your tool came to mind. 😄
